src/main/java/au/org/ala/names/util/DwcaWriter.java

```diff
@@ -0,0 +1,413 @@
+     * Add a single value for the current core record.
+     * Calling this method requires that #newRecord() has been called at least once,
+     * otherwise an IllegalStateException is thrown.
+     * @param term
+     * @param value
+     */
+    public void addCoreColumn(Term term, String value) {
+        // ensure we do not overwrite the coreIdTerm if one is defined
+        if (coreIdTerm != null && coreIdTerm.equals(term)) {
+            throw new IllegalStateException("You cannot add a term that was specified as coreId term");
+        }
+
+        List<Term> coreTerms = terms.get(coreRowType);
+        if (!coreTerms.contains(term)) {
+            if (useHeaders && recordNum>1){
+                throw new IllegalStateException("You cannot add new terms after the first row when headers are enabled");
+            }
+            coreTerms.add(term);
+        }
+        try {
+            coreRow.put(term, value);
+        } catch (NullPointerException e) {
+            // no core record has been started yet
+            throw new IllegalStateException("No core record has been created yet. Call newRecord() at least once");
+        }
+    }
+
+    /**
+     * Add a default value to a term of the core.
+     *
+     * @param term
+     * @param defaultValue
+     */
+    protected void addCoreDefaultValue(Term term, String defaultValue){
+        addDefaultValue(coreRowType, term, defaultValue);
+    }
+
+    /**
+     * Add a default value to a term of the provided rowType.
+     *
+     * @param rowType
+     * @param term
+     * @param defaultValue
+     */
+    public void addDefaultValue(Term rowType, Term term, String defaultValue){
+
+        if(!defaultValues.containsKey(rowType)){
+            defaultValues.put(rowType, new HashMap<Term, String>());
+        }
+        Map<Term,String> currentDefaultValues= defaultValues.get(rowType);
+        if(currentDefaultValues.containsKey(term)){
+            throw new IllegalStateException("The default value of term "+ term + " is already defined");
+        }
+        currentDefaultValues.put(term, defaultValue);
+    }
+
+    /**
+     * @return new map of all current data file names by their rowTypes.
+     */
+    public Map<Term, String> getDataFiles() {
+        return Maps.newHashMap(dataFileNames);
+    }
+
+    protected void addExtensionRecord(Term rowType, Map<Term, String> row, boolean detatched) throws IOException {
+        // make sure we know the extension rowtype
+        if (!terms.containsKey(rowType)) {
+            addRowType(rowType, detatched);
+        }
+
+        // make sure we know all terms
+        List<Term> knownTerms = terms.get(rowType);
+        final boolean isFirst = knownTerms.isEmpty();
+        for (Term term : row.keySet()) {
+            if (!knownTerms.contains(term)) {
+                if (useHeaders && !isFirst){
+                    throw new IllegalStateException("You cannot add new terms after the first row when headers are enabled");
```

Comment:
New terms shouldn't ever be allowed after the first row of the file, as it complicates processing with many CSV files that (understandably) complain about row lengths being different.